### DISCUSSION TITLE
pynecone 套件介紹

### DISCUSSION THREAD

**Discussion by quan0715** [1101-GS4719-C/Python-Community-GS4719-C#1305] [優秀Python套件]
# pynecone 套件介紹
### 簡介
一個全新的python網頁 「全端框架」，主打兩項特點
1. 快速開發，適用各種use case
    * 全端框架用 python 搞定，不需要額外撰寫API去串聯你的前後端。
    * 精美的UI Kit, 現成的套件搞定一切，不需要花費過多心力也可以美化你的個人網站
2. 全python 框架
    * 用 class 表示狀態而class 裡面的method 更新狀態，而狀態再去渲染整個UI，使用流程符合OOP架構設計，容易上手
   
更多細節請看[github 專案](https://github.com/pynecone-io/pynecone) 或官方網站[pynecone](https://pynecone.io/)

### 小心得
使用起來相對於其他網頁框架，使用起來超好上手，受惠於全端框架的部分，從UI到Database到deploy你的app這個框架都可以一條龍完成。你也可以去套用其他現行的python 套件，例如 pandas/numpy 等等幫助你完成你的 data science dashboard project 等等，雖然目前還在開發階段，但蠻有潛力發展成主流框架系統的，大家可以試玩看看，說不定他可以幫你快速開發並展示你的side project.

### 小小使用程式碼
這是一個簡單的計數器程式碼
<img width="730" alt="image" src="https://user-images.githubusercontent.com/60366187/235454392-2088b9cd-75d2-4e49-9d8c-d72b1cd888fd.png">

```python3
import pynecone as pc
# import 此套件

# 用class 設定 state (狀態)
class State(pc.State):
    count: int = 0

    def increment(self):
        # 加事件
        self.count += 1

    def decrement(self):
        # 減事件
        self.count -= 1

# 網頁主體，利用嵌套（俄羅斯娃娃套娃的方式），來去表示UI
# 以此為例 stack (button, heading, button)
def index():
    return pc.hstack(
        pc.button(
            "Decrement",
            color_scheme="red",
            border_radius="1em",
            on_click=State.decrement,
        ),
        pc.heading(State.count, font_size="2em"),
        pc.button(
            "Increment",
            color_scheme="green",
            border_radius="1em",
            on_click=State.increment,
        ),
    )


app = pc.App(state=State) # 定義app主體，賦予狀態
app.add_page(index) # 加入 page
app.compile() # 編譯、執行
```
這樣就完成了簡單的計時器了，你學廢了嗎，快去玩玩看吧

**Comment by ziihui** [comment]
學廢嚕

**Comment by Jack-1D** [comment]
好強

**Reply by quan0715** [reply]
歡迎使用此框架來自己製作，你自己的期末報告 ==
